punchcut/src/source/glyf/deltas.rs

```diff
@@ -0,0 +1,289 @@
+    out_points: &mut [Point<Fixed>],
+) -> Option<()> {
+    let ref_in = points.get(ref_)?.map(Fixed::from_i32);
+    let ref_out = out_points.get(ref_)?;
+    let delta = *ref_out - ref_in;
+    if delta.x == Fixed::ZERO && delta.y == Fixed::ZERO {
+        return Some(());
+    }
+    for out_point in out_points.get_mut(start..ref_)? {
+        *out_point += delta;
+    }
+    for out_point in out_points.get_mut(ref_ + 1..=end)? {
+        *out_point += delta;
+    }
+    Some(())
+}
+
+/// Interpolate the coordinates of all points between `start` and `end` using
+/// `ref1` and `ref2` as the reference point indices.
+///
+/// Modeled after the FreeType implementation: <https://github.com/freetype/freetype/blob/bbfcd79eacb4985d4b68783565f4b494aa64516b/src/truetype/ttgxvar.c#L3813>
+///
+/// For details on the algorithm, see: <https://learn.microsoft.com/en-us/typography/opentype/spec/gvar#inferred-deltas-for-un-referenced-point-numbers>
+fn interpolate_range(
+    points: &[Point<i32>],
+    start: usize,
+    end: usize,
+    ref1: usize,
+    ref2: usize,
+    out_points: &mut [Point<Fixed>],
+) -> Option<()> {
+    if start > end {
+        return Some(());
+    }
+    // FreeType uses pointer tricks to handle x and y coords with a single piece of code.
+    // Try a macro instead.
+    macro_rules! interp_coord {
+        ($coord:ident) => {
+            let mut ref1 = ref1;
+            let mut ref2 = ref2;
+            if points.get(ref1)?.$coord > points.get(ref2)?.$coord {
+                core::mem::swap(&mut ref1, &mut ref2);
+            }
+            let in1 = Fixed::from_i32(points.get(ref1)?.$coord);
+            let in2 = Fixed::from_i32(points.get(ref2)?.$coord);
+            let out1 = out_points.get(ref1)?.$coord;
+            let out2 = out_points.get(ref2)?.$coord;
+            let d1 = out1 - in1;
+            let d2 = out2 - in2;
+            // If the reference points have the same coordinate but different delta,
+            // inferred delta is zero. Otherwise interpolate.
+            if in1 != in2 || out1 == out2 {
+                let scale = if in1 != in2 {
+                    (out2 - out1) / (in2 - in1)
+                } else {
+                    Fixed::ZERO
+                };
+                let range = start..=end;
+                for (point, out_point) in points
+                    .get(range.clone())?
+                    .iter()
+                    .zip(out_points.get_mut(range.clone())?)
+                {
+                    let mut out = Fixed::from_i32(point.$coord);
+                    if out <= in1 {
+                        out += d1;
+                    } else if out >= in2 {
+                        out += d2;
+                    } else {
+                        out = out1 + (out - in1) * scale;
+                    }
+                    out_point.$coord = out;
+                }
+            }
+        };
+    }
+    interp_coord!(x);
+    interp_coord!(y);
+    Some(())
+}
```

Comment:
Computation of deltas, interpolating ranges, shifting ranges, etc all seem remarkably testable?

Reply:
Passing a function to get variation data instead of an entire Gvar might make testing easier?

Reply:
Testing full delta computation in isolation of glyf is annoying because the algorithm depends on the outline to determine interpolation factors for deltas. I did, however, add unit tests for the underlying delta interpolation code which is designed to trigger both interpolations and shifts.

I think this, combined with the local tests in gvar plus the data driven tests gives us good coverage. Happy to extend this in another PR if you feel it's insufficient.